server/src/plugins/queryFilter.js

```diff
@@ -0,0 +1,142 @@
+const Boom = require('@hapi/boom')
+const Bounce = require('@hapi/bounce')
+const Joi = require('@hapi/joi')
+const { Filters } = require('../utils/Filter')
+const { parseFilterString } = require('../utils/filterUtils')
+
+/**
+ * This plugin hooks into onPrehandler, processing all requests it's enabled for.
+ * It's designed to be used in conjunction with CustomJoi.js that extends Joi
+ * with a filter type. The filters are validated through Hapi's validate.query.
+ *
+ * The purpose of the plugin is to group all filters in `request.plugins.queryFilter.filters`.
+ * The value of this property is an object of Filters.
+ *
+ * All options can be overwritten by each route.
+ * Options:
+ *  enabled - enable the queryFilter
+ *  ignoreKeys - ignore the key. Note that only keys of type Joi.filter() in the validate.query object are used.
+ *  rename - A Joi-schema with .rename() functions that are used to get renames of the keys.
+ *           Can also be a plain Object with keys being the query-name, and the value being the database-name.
+ *      This is used to rename API-facing filter-names to database-column names.
+ */
+
+const FILTER_TYPE = 'filter'
+
+const defaultOptions = {
+    enabled: false, // default disabled for all routes
+    method: ['GET'],
+    ignoreKeys: ['paging'],
+}
+
+// Holds Joi-schema descriptions, as these can be quite expensive to generate and this is a hot-path
+// cached by path as key
+
+// validateDescriptions is used to get queryParams with Filter-type
+const validateDescriptions = {}
+// renameDescriptions is used to rename the filter-Keys to the correct DB-column
+const renameDescriptions = {}
+
+const onPreHandler = function(request, h) {
+    const routeOptions = request.route.settings.plugins.queryFilter || {}
+
+    const options = {
+        ...this.options,
+        ...routeOptions,
+        ignoreKeys: this.options.ignoreKeys.concat(
+            routeOptions.ignoreKeys || []
+        ),
+    }
+
+    if (
+        !options.enabled ||
+        !options.method.includes(request.method.toUpperCase())
+    ) {
+        return h.continue
+    }
+
+    const routeQueryValidation = request.route.settings.validate.query
+    const rename = routeOptions.rename
+    const queryFilterKeys = new Set()
+    let renameMap = rename
+    let validateDescription = validateDescriptions[request.path]
+    let renameDescription = renameDescriptions[request.path]
+
+    if (rename) {
+        if (Joi.isSchema(rename)) {
+            if (!renameDescription) {
```

Comment:
personally I'd add these to the same if-case with && in between to avoid nesting scopes so deep as there are no else cases